### PR TITLE
Update NodeWarningProperties GetIntersect() function

### LINKED ISSUE #11669: [Bug]: Very slow restore or OOM when using NoWarn
### NuGet Product Used

dotnet.exe, NuGet.exe

### Product Version

6.1.0.106

### Worked before?

_No response_

### Impact

I'm unable to use this version

### Description

A NoWarn in a package reference can add minutes to a restore, or make it fail with an `OutOfMemoryException`.

This is related to https://github.com/NuGet/Home/issues/11222 and https://github.com/NuGet/NuGet.Client/pull/4475.

### Background

`TransitiveNoWarnUtils` performs a breadth-first traversal of the dependency graph.

The first time we find a path from the root node to a project node `v`, we store the union of the NoWarns seen along the path in a dictionary called `seen`.

Each subsequent time we find a path from the root node to `v`, we check if the NoWarns seen along the new path to `v` are a superset of `seen[v]`.

If that's the case, we don't need to process the node again.

If that's _not_ the case, we update `seen[v]` to be the intersection of `seen[v]` and the NoWarns seen along the new path to `v`.

There is a bug in the method for finding the intersection that causes nodes to be processed more often than they should be. As a consequence, a single NoWarn can add minutes to a project restore, or cause the restore to fail with an `OutOfMemoryException`.

In the method for checking whether the NoWarns along one path are a superset of the NoWarns along another path (`NodeWarningProperties.IsSubSetOf`), `null` is considered to be the empty set.

The bug is this: in the method to get the intersection of the NoWarns (`NodeWarningProperties.GetIntersect`), `null` doesn't act like the empty set.

### Steps to reproduce

To reproduce this issue, create n >= 30 projects - where `ClassLibrary{i}` has project references to `ClassLibary1`..`ClassLibrary{i-1}`.

In addition, create three more projects: `Root`, `RootLeft`, and `RootRight` with the following project references:

```
Root -> RootLeft -> ClassLibrary{n}
Root -> RootRight -> ClassLibrary{n}
```

Add the following two package references to `RootLeft`:

    <PackageReference Include="Microsoft.EntityFrameworkCore" Version="6.0.0">
      <NoWarn>NU1000</NoWarn>
    </PackageReference>
    <PackageReference Include="Microsoft.Extensions.DependencyInjection" Version="5.0.2" />

(Code to create all these projects can be found here: https://github.com/mjolka/LotsOfDependenciesIntersect)

The single nowarn in `RootLeft` causes the package restore to go from ~3 seconds to ~7 minutes when n = 30.

When n = 35, the restore fails with an `OutOfMemoryException`.

The reason being, all `ClassLibrary{i}` projects are found first via a path that goes through `RootLeft`. This path contains a single NoWarn.

When `ClassLibrary{i}` is later found via a path that goes through `RootRight`, these are no NoWarns along the path. The code to calculate the intersection of the NoWarns should result in the empty set. However, because of the bug, after taking the intersection, `seen[v]` still contains a single NoWarn, and the node will be processed again each time we find a new path to it.

The fix is to align the behaviour of `GetIntersect` to that of `IsSubSetOf` by treating `null` as the empty set. I'm able to contribute a pull request with this change.

### Verbose Logs

_No response_

### PR DESCRIPTION
**Pull request by wrall**
## Bug

Fixes: https://github.com/NuGet/Home/issues/11669

Regression? Last working version:
No

## Description
Aligns the behavior of `GetIntersect` to that of `IsSubSetOf` by treating `null` as the empty set, and using null in favor of empty sets and empty dictionaries.
## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] Automated tests added
  - **OR**
  - [ ] Test exception
  - **OR**
  - [ ] N/A

- **Documentation**
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

### CODE REVIEW

**Comment by ghost** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/NuGet/NuGet.Client?pullRequest=4644) <br/>All CLA requirements met.

**Comment by nkolev92** [comment]
@wrall 

Thanks for the contribution! :)

For reference, you're probably best not using `[]` in the issue link as GH won't automatically detect and close the issue. 
I've fixed the description for now. 

Please check the respective boxes of the template.

**Review by nkolev92 — COMMENTED** [review]
This looks great. 

Based on some of the changes, it seems like there should be a functional change here as well no? 
Can we add a test for that.

**Comment by erdembayar** [comment]
Thank you for contribution.
Based original issue it looks main purpose of this PR is perf improvement. Do we have before and after memory allocation measurement? If yes, then what is the improvement?

**Comment by mjolka** [comment]
Hi @erdembayar,

I can field that question. The perf improvement isn't about memory allocations, but allowing the algorithm to stop processing a node once we have found a path to the node with no NoWarns. (https://github.com/NuGet/Home/wiki/%5BSpec%5D-Transitive-Warning-Properties#implementation-details)

I've attached a solution that can be used to reproduce the issue: [LotsOfDependencies.zip](https://github.com/NuGet/NuGet.Client/files/8790687/LotsOfDependencies.zip). The problem is that we keep processing nodes due to the bug in taking the intersection, even though we have found a path to the node with no NoWarns.

As for numbers, before wrall's change (built from commit 9e244eb):

```
Task Performance Summary:
        0 ms  CheckIfPackageReferenceShouldBeFrameworkReference   2 calls
        0 ms  ApplyImplicitVersions                      1 calls
        2 ms  NuGetMessageTask                           1 calls
        4 ms  WarnForInvalidProjectsTask                 1 calls
        6 ms  GetRestoreSolutionProjectsTask             1 calls
        7 ms  Message                                   68 calls
        9 ms  CheckForDuplicateFrameworkReferences      33 calls
       18 ms  ResolveAppHosts                           33 calls
       26 ms  CheckForImplicitPackageReferenceOverrides  33 calls
       28 ms  GetRestorePackageDownloadsTask            33 calls
       28 ms  RemoveDuplicates                          69 calls
       30 ms  GetRestoreFrameworkReferencesTask         33 calls
       33 ms  GetProjectTargetFrameworksTask            33 calls
       37 ms  CheckForDuplicateNuGetItemsTask           99 calls
       50 ms  ConvertToAbsolutePath                     66 calls
       53 ms  GetRestorePackageReferencesTask           33 calls
       56 ms  GetRestoreProjectReferencesTask           33 calls
       79 ms  GetRestoreProjectStyleTask                33 calls
       94 ms  GetRestoreDotnetCliToolsTask              33 calls
      592 ms  ProcessFrameworkReferences                33 calls
     1476 ms  GetRestoreSettingsTask                    33 calls
     7286 ms  MSBuild                                   38 calls
    357163 ms  RestoreTask                                1 calls
```

After wrall's change:

```
Task Performance Summary:
        0 ms  CheckIfPackageReferenceShouldBeFrameworkReference   2 calls
        0 ms  ApplyImplicitVersions                      1 calls
        2 ms  NuGetMessageTask                           1 calls
        2 ms  WarnForInvalidProjectsTask                 1 calls
        4 ms  CheckForDuplicateFrameworkReferences      33 calls
        4 ms  Message                                   68 calls
        5 ms  CheckForImplicitPackageReferenceOverrides  33 calls
        6 ms  GetRestorePackageDownloadsTask            33 calls
        6 ms  GetRestoreSolutionProjectsTask             1 calls
        7 ms  GetRestoreFrameworkReferencesTask         33 calls
        8 ms  GetProjectTargetFrameworksTask            33 calls
        8 ms  GetRestorePackageReferencesTask           33 calls
        9 ms  RemoveDuplicates                          69 calls
        9 ms  CheckForDuplicateNuGetItemsTask           99 calls
       10 ms  ResolveAppHosts                           33 calls
       11 ms  GetRestoreDotnetCliToolsTask              33 calls
       12 ms  ConvertToAbsolutePath                     66 calls
       14 ms  GetRestoreProjectStyleTask                33 calls
       23 ms  GetRestoreProjectReferencesTask           33 calls
      117 ms  ProcessFrameworkReferences                33 calls
      476 ms  GetRestoreSettingsTask                    33 calls
     1482 ms  RestoreTask                                1 calls
     4003 ms  MSBuild                                   38 calls
```
 
@wrall thank you for the pull request!

**Comment by mjolka** [comment]
Apologies, I forgot that due to the extra processing of nodes, there was in fact an issue with memory as well.

I've attached screenshots from dotMemory, using the same solution as in my previous comment.

Before:

dotMemory UI 64_3kT9TV5FHX

After:

6hr3gv7zRg

**Comment by wrall** [comment]
Thanks @mjolka for those numbers!  Unfortunately I've planned poorly and submitted this just before leaving for a week.  I'll try to pick this up late next week to gather more information, add a functional test, and possibly make another performance change (specifically: use a shared "Empty" NodeWarningProperties object and make it immutable, as I think that will be the most common in very-large projects).

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs [hunk 1]
```diff
@@ -809,11 +811,21 @@ public NodeWarningProperties GetIntersect(NodeWarningProperties other)
                         thisPackages.TryGetValue(key, out var thisCodes);
```
Comment (mjolka) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs]: There's an opportunity to improve the existing logic. Since this is an intersection, we don't need `allKeys` - we can iterate over just the keys/values in `thisPackages`.

Comment (wrall) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs]: Excellent point!

**Comment by wrall** [comment]
With the latest change, I see a nice reduction in the number of NodeWarningProperties objects allocated for a sample solution (from 673 to 73).  I couldn't come up with a functional test that showed any difference in behavior at a higher level.  But given the fact that we are just following the behavior in the spec that was written for this originally (https://github.com/NuGet/Home/wiki/%5BSpec%5D-Transitive-Warning-Properties#implementation-details), I don't have a strong worry about this change.

@erdembayar, please review when you have a chance.

**Comment by erdembayar** [comment]
> With the latest change, I see a nice reduction in the number of NodeWarningProperties objects allocated for a sample solution (from 673 to 73). I couldn't come up with a functional test that showed any difference in behavior at a higher level. But given the fact that we are just following the behavior in the spec that was written for this originally (https://github.com/NuGet/Home/wiki/%5BSpec%5D-Transitive-Warning-Properties#implementation-details), I don't have a strong worry about this change.
> 
> @erdembayar, please review when you have a chance.

Thank you for latest commit. I'll try to review this week, meantime could you be able to measure memory allocation screenshot from either with VS memory profiler or dotmemory? I want to see actual performance improvement before start debugging for correctness aspect.

File: src/NuGet.Core/NuGet.Commands/PublicAPI/net472/PublicAPI.Shipped.txt [hunk 2]
```diff
@@ -858,18 +858,16 @@ NuGet.Commands.ToolRestoreUtility
 NuGet.Commands.TransitiveNoWarnUtils
 NuGet.Commands.TransitiveNoWarnUtils.DependencyNode
 NuGet.Commands.TransitiveNoWarnUtils.DependencyNode.DependencyNode(string id, bool isProject, NuGet.Commands.TransitiveNoWarnUtils.NodeWarningProperties nodeWarningProperties) -> void
-NuGet.Commands.TransitiveNoWarnUtils.DependencyNode.DependencyNode(string id, bool isProject, System.Collections.Generic.HashSet<NuGet.Common.NuGetLogCode> projectWideNoWarn, System.Collections.Generic.Dictionary<string, System.Collections.Generic.HashSet<NuGet.Common.NuGetLogCode>> packageSpecificNoWarn) -> void
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/PublicAPI/net472/PublicAPI.Shipped.txt]: We try not to change Shipped APIs in minor versions. 

Please refer to https://github.com/NuGet/NuGet.Client/blob/dev/docs/nuget-sdk.md#api-compatibility-policy to learn how we handle new api changes.

**Review by nkolev92 — COMMENTED** [review]
>  I couldn't come up with a functional test that showed any difference in behavior at a higher level

Interesting. I'm surprised that's the case.

Please address the public API changes and we'll take another looks. 
Thanks!

File: src/NuGet.Core/NuGet.Commands/PublicAPI/net472/PublicAPI.Shipped.txt (hunk 2 above)
Comment (wrall) [reply src/NuGet.Core/NuGet.Commands/PublicAPI/net472/PublicAPI.Shipped.txt]: Thanks for the link.  I hadn't realized that external systems could be using these internal pieces.  I'll try to go with an approach that should be compatible.

**Comment by wrall** [comment]
> Thank you for latest commit. I'll try to review this week, meantime could you be able to measure memory allocation screenshot from either with VS memory profiler or dotmemory? I want to see actual performance improvement before start debugging for correctness aspect.

I've run restore using both the current "dev" branch and my current "dev-wrall-NodeWarningProperties-intersect" branch on a very-large set of projects that I have and have profiled the allocations using dotMemory.  Sadly I can't share the set of projects publicly - feel free to reach out to me internally if you would like more information.

First, the full msbuild restore takes ~8 minutes using "dev" and ~7 minutes using "dev-wrall-NodeWarningProperties-intersect".

Simple (sampled) memory profile isn't very exciting...
dev:
restore unchanged memory

dev-wrall-NodeWarningProperties-intersect:
restore changed memory

What's really interesting is the difference in allocations when using full collection mode.

dev:
restore unchanged allocations
restore unchanged allocation_callstacks

dev-wrall-NodeWarningProperties-intersect:
restore changed allocations
restore changed allocation_callstacks

As you can see, ~20% of all allocations were happening during NodeWarningProperties.GetIntersect(), and that has been cut down to about 0.1%.

File: src/NuGet.Core/NuGet.Commands/PublicAPI/net472/PublicAPI.Shipped.txt (hunk 2 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/PublicAPI/net472/PublicAPI.Shipped.txt]: I think it's unlikely that many people use this piece of code, but it's easier to be consistent as we don't really have any data on what APIs people actually do use.

File: src/NuGet.Core/NuGet.Commands/PublicAPI/net472/PublicAPI.Unshipped.txt [hunk 3]
```diff
@@ -0,0 +1,4 @@
+NuGet.Commands.TransitiveNoWarnUtils.NodeWarningProperties.ExtractPathNoWarnProperties(string libraryId) -> System.Collections.Generic.HashSet<NuGet.Common.NuGetLogCode>
+NuGet.Commands.TransitiveNoWarnUtils.NodeWarningProperties.Merge(NuGet.Commands.TransitiveNoWarnUtils.NodeWarningProperties other) -> NuGet.Commands.TransitiveNoWarnUtils.NodeWarningProperties
+static NuGet.Commands.TransitiveNoWarnUtils.NodeWarningProperties.Create(System.Collections.Generic.HashSet<NuGet.Common.NuGetLogCode> projectWide, System.Collections.Generic.Dictionary<string, System.Collections.Generic.HashSet<NuGet.Common.NuGetLogCode>> packageSpecific) -> NuGet.Commands.TransitiveNoWarnUtils.NodeWarningProperties
+static readonly NuGet.Commands.TransitiveNoWarnUtils.NodeWarningProperties.Empty -> NuGet.Commands.TransitiveNoWarnUtils.NodeWarningProperties
```
Comment (zivkan) [inline src/NuGet.Core/NuGet.Commands/PublicAPI/net472/PublicAPI.Unshipped.txt]: All of these APIs appear to be called from within the same assembly. Can we make all these APIs internal?

Customers using NuGet.Commands in their own app probably don't have a reason to use these APIs.

Comment (wrall) [reply src/NuGet.Core/NuGet.Commands/PublicAPI/net472/PublicAPI.Unshipped.txt]: Sure.  Going forward, it would make sense to make most of the APIs in this file internal.  Maybe everything except the CreateTransitiveWarningPropertiesCollection call.  I understand that would require a major version change.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs [hunk 4]
```diff
@@ -720,37 +689,44 @@ private class LookUpNode
...
 
-
         /// <summary>
-        /// A class to hold minimal version of project wide nowarn and package specific no warn for a project.
+        /// An immutable class to hold minimal version of project wide nowarn and package specific no warn for a project.
         /// </summary>
         public class NodeWarningProperties : IEquatable<NodeWarningProperties>
         {
+            // Empty NodeWarningProperties singleton - to avoid allocation for most-common case
+            internal static readonly NodeWarningProperties Empty = new NodeWarningProperties(null, null);
```
Comment (jeffkl) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs:701]: I like this optimization so that projects with no suppressed warnings don't pay a penalty.  I'm curious if we could cache more.  For example, I've suppressed a warning or set of warnings in `Directory.Build.props` so every project has the exact same set.  Can we use the same object for each project in that case?  Essentially we'd only allocate a new instance if a project had a different set?  We could of course do this in a future change but wanted to throw it out there.  If we could quickly look up that an immutable instance has already been created that represents the warnings, we could just return that instead of allocating a new one.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs [hunk 5]
```diff
@@ -720,37 +689,44 @@ private class LookUpNode
...
+
             // ProjectWide NoWarn properties
-            public HashSet<NuGetLogCode> ProjectWide { get; }
+            private readonly HashSet<NuGetLogCode> _projectWide;
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs]: Why are we doing this refactoring?
I'm not sure what the functional benefit here is. 

Making a lot of changes makes it really difficult to keep up with the actual relevant changes.

Comment (wrall) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs]: I apologize that the review has grown uglier.  My goal here was to emphasize that this HashSet and the below Dictionary are intended to be immutable.  Note that the "{ get; }" syntax is a shorthand for a getter property with a backing member variable that can only be set by the constructor.  This is basically equivalent to a private readonly variable with a getter property that returns it.

Originally, I had planned to remove the getter property altogether.  Unfortunately, this object had made it into the publicly-exposed API for NuGet, so we cannot make that change without a major version update.  I decided to keep it with the separated readonly member variable to hopefully help remind the next updater that this is something that can and should be removed at the next major version update.

I can revert this if you prefer and go back to the old syntax with a comment.

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs]: As you mentioned, they are semantically equivalent, so it's probably easier to review if we didn't make those changes.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs (hunk 4 above)
Comment (wrall) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs:701]: That's certainly an interesting idea!  I don't have a good approach in mind for caching the dictionary or set though - perhaps if we had the knowledge here about the warnings coming from a shared set.

Perhaps another way of handling that situation would be to add something to GetIntersect that checks if the two nodes are equal, and if so then return one.  Or, if just the hash sets or dictionaries are equal, just reuse it.  Right now it handles the ReferenceEqual scenario, but not value-equal.  Adding a value-equal check would optimize Memory at a cost to CPU.  But perhaps we could also fit this into the IsSubset check - perhaps that could have a special case for equals?

This could be an interesting thing to file a bug about - performance for common suppressed warning(s) scenario.

Comment (wrall) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs:701]: I've filed https://github.com/NuGet/Home/issues/11878 to track this idea.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs [hunk 6]
```diff
@@ -658,17 +627,17 @@ public class DependencyNode : IEquatable<DependencyNode>
             // If a node is a project then it will hold these properties
             public NodeWarningProperties NodeWarningProperties { get; }
 
-            public DependencyNode(string id, bool isProject, HashSet<NuGetLogCode> projectWideNoWarn, Dictionary<string, HashSet<NuGetLogCode>> packageSpecificNoWarn)
+            public DependencyNode(string id, bool isProject, NodeWarningProperties nodeWarningProperties)
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs]: somehow this constructor and the one below got flipped. No functional issues, just the order of them flipped creating a bigger changset.

**Review by nkolev92 — DISMISSED** [review]
Thanks for the contribution!

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs (hunk 6 above)
Comment (wrall) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs]: Ah, my apologies - I was responding to PR feedback about not changing the public APIs and must have pasted it back in the wrong order.  Thank you for the review!

**Comment by wrall** [comment]
> Thanks for the contribution!

Could you re-add the Approved for CI tag?  It looks like @jeffkl removed it and forgot to add it back.

**Comment by jeffkl** [comment]
Sorry about that, I swear I removed and readded it!

**Comment by wrall** [comment]
Here's the performance change in a handy chart, based on something my team restores frequently.

| Change | Objects allocated | Bytes allocated | Objects % | Bytes % |
| ------ | ----------------- | --------------- | --------- | ------- |
| dev (a92decd2847a776c0ead2c1658f31f5725900ab3) | 1,529.41M objects | 89.74 GB | 100% | 100% |
| dev + #4644 | 1,235.15M objects | 68.08 GB | 80.76% | 75.86% |
| dev + #4644 + #4666 | 997.81M objects | 61.14 GB | 65.24% | 68.13% |

Note that the performance becomes far worse when the projects are all converted from referencing each other using a mix of binary and project references to solely using project references, with the "before" scenario becoming a DNF (making it hard to compare).

**Comment by erdembayar** [comment]
> Here's the performance change in a handy chart, based on something my team restores frequently.
> 
> Change	Objects allocated	Bytes allocated	Objects %	Bytes %
> dev ([a92decd](https://github.com/NuGet/NuGet.Client/commit/a92decd2847a776c0ead2c1658f31f5725900ab3))	1,529.41M objects	89.74 GB	100%	100%
> dev + #4644	1,235.15M objects	68.08 GB	80.76%	75.86%
> dev + #4644 + #4666	997.81M objects	61.14 GB	65.24%	68.13%
> Note that the performance becomes far worse when the projects are all converted from referencing each other using a mix of binary and project references to solely using project references, with the "before" scenario becoming a DNF (making it hard to compare).

Thank you for posting delta info. I'll review tomorrow, I was busy with other tasks recently.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs [hunk 7]
```diff
@@ -173,34 +169,23 @@ public static class TransitiveNoWarnUtils
...
-                        // Merge the node's package specific no warn to the one in the path.
-                        var mergedPackageSpecificNoWarn = MergePackageSpecificNoWarn(pathPackageSpecificNoWarn, nodePackageSpecificNoWarn);
-
+                        // Merge the node's no warn properties with the one in the path.
+                        var nodeWarningProperties = nodeLookUp.NodeWarningProperties;
                         AddDependenciesToQueue(nodeDependencies,
                             queue,
-                            mergedProjectWideNoWarn,
-                            mergedPackageSpecificNoWarn);
+                            nodeWarningProperties.Merge(node.NodeWarningProperties));
```
Comment (erdembayar) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs:182]: Do we need to create new method `Merge` if it's only referenced once?
If we have them here in place, then it's much easier to reason with it.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs [hunk 8]
```diff
@@ -173,34 +169,23 @@ public static class TransitiveNoWarnUtils
...
                     if (nodeIsProject)
                     {
-                        // Merge the node's project wide no warn to the one in the path.
-                        var mergedProjectWideNoWarn = MergeCodes(pathProjectWideNoWarn, nodeProjectWideNoWarn);
-
-                        // Merge the node's package specific no warn to the one in the path.
-                        var mergedPackageSpecificNoWarn = MergePackageSpecificNoWarn(pathPackageSpecificNoWarn, nodePackageSpecificNoWarn);
-
+                        // Merge the node's no warn properties with the one in the path.
+                        var nodeWarningProperties = nodeLookUp.NodeWarningProperties;
```
Comment (erdembayar) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs]:
```suggestion
                        NodeWarningProperties nodeWarningProperties = nodeLookUp.NodeWarningProperties;
```

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs [hunk 9]
```diff
@@ -335,20 +317,7 @@ private static PackageSpec GetNodePackageSpec(LocalMatch localMatch)
...
-
-            if (nodeWarningProperties?.PackageSpecific?.Count > 0 &&
-                nodeWarningProperties.PackageSpecific.TryGetValue(libraryId, out var codes) &&
-                codes?.Count > 0)
-            {
-                result.UnionWith(codes);
-            }
-
-            return result;
+            return nodeWarningProperties.ExtractPathNoWarnProperties(libraryId);
```
Comment (erdembayar) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs:321]: Do we really need this change? This method `ExtractPathNoWarnProperties` not referenced by any other place. Could we simply remove it?

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs (hunk 6 above)
Comment (erdembayar) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs]: It looks issue is still persist even though comment is resolved. I'm changing back to unresolved state.

File: test/NuGet.Core.Tests/NuGet.Commands.Test/TransitiveNoWarnUtilsTests.cs [hunk 10]
```diff
@@ -1173,6 +1158,318 @@ public void NodeWarningPropertiesIsSubSetOf_PackageSpecificNoWarnsHaveDifferentK
...
+        [InlineData(new object[] { new[] { "test_id1:NU1605, NU1604, NU1701" }, new[] { "test_id1:NU1703, NU1605, NU1604" }, new[] { "test_id1:NU1605, NU1604" } })]
+        [InlineData(new object[] { new[] { "test_id1:NU1605, NU1604, NU1701", "test_id2:NU1703, NU3018, NU1701" }, new[] { "test_id1:NU1703, NU3018, NU1701", "test_id2:NU1604" }, new[] { "test_id1:NU1701" } })]
+        [InlineData(new object[] { new[] { "test_id1:NU1605, NU1604, NU1701", "test_id3:NU1703, NU3018, NU1701" }, new[] { "test_id2:NU1703, NU3018, NU1701", "test_id3:NU1604, NU3018, NU1701" }, new[] { "test_id3:NU3018, NU1701" } })]
+        public void NodeWarningPropertiesGetIntersect_PackageSpecificIntersecting(
+            string[] firstPackageSpecificNoWarn,
+            string[] secondPackageSpecificNoWarn,
+            string[] expectedPackageSpecificNoWarn)
+        {
+            // Arrange
+            var firstNode = TransitiveNoWarnUtils.NodeWarningProperties.Create(
```
Comment (erdembayar) [inline test/NuGet.Core.Tests/NuGet.Commands.Test/TransitiveNoWarnUtilsTests.cs:1454]: Explicit type "NodeWarningProperties" please, other instances below too.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs [hunk 11]
```diff
@@ -777,43 +748,54 @@ public bool Equals(NodeWarningProperties other)
...
+                Dictionary<string, HashSet<NuGetLogCode>> packages = null;
+                if (thisPackages != null && otherPackages != null)
                 {
                     packages = new Dictionary<string, HashSet<NuGetLogCode>>(StringComparer.OrdinalIgnoreCase);
 
-                    var allKeys = thisPackages.Keys.Concat(otherPackages.Keys)
-                        .Distinct(StringComparer.OrdinalIgnoreCase);
-
-                    foreach (var key in allKeys)
+                    foreach (var pair in thisPackages)
```
Comment (erdembayar) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs]: Please use explicit types instead of `var`.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs [hunk 12]
```diff
@@ -902,18 +934,23 @@ private static HashSet<NuGetLogCode> Intersect(HashSet<NuGetLogCode> first, Hash
                     return first;
                 }
 
-                if (first == null)
+                // null is empty and cannot intersect
+                if (first == null || first.Count == 0)
                 {
-                    return second;
+                    return null;
```
Comment (erdembayar) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs:941]: Are we breaking any edge case previously working but now it doesn't work anymore?
I'm not familiar with this part of the code, so can't think of one for now.

**Review by erdembayar — COMMENTED** [review]
Let's start with some comments, in general looks good.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs (hunk 7 above)
Comment (wrall) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs:182]: I think that it is far safer design-wise to reduce the number of things that are accessing the innards of the NodeWarningProperties class if we can avoid it.  In the perfect world, we would remove the ability for the code here to access the Hashset/Dictionary directly.  I'd prefer to keep it in an instance method for the sake of separation of concerns.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs (hunk 9 above)
Comment (wrall) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs:321]: Same thing here - trying to reduce the things that reach inside of NodeWarningProperties.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs (hunk 11 above)
Comment (wrall) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs]: Sorry.  Trying to follow https://github.com/NuGet/NuGet.Client/blob/dev/docs/coding-guidelines.md rule 10, but I personally avoid var all of the time so it's hard to find the balance.  Also, I think that this one pre-dated my change and I filed to update it.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs (hunk 12 above)
Comment (wrall) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs:941]: I couldn't find any edge-case that was broken.  But I'll admit that I'm coming to this trying to tackle a performance blocker and am not an expert on NuGet warnings.

**Review by erdembayar — APPROVED** [review]
Great contribution, thank you.

**Review by nkolev92 — APPROVED** [review]

**Comment by wrall** [comment]
@erdembayar, let me know if you need anything else before merge.
